evm_arithmetization/src/prover.rs

```diff
@@ -114,8 +116,14 @@ where
         .map(|c| c.merkle_tree.cap.clone())
         .collect::<Vec<_>>();
     let mut challenger = Challenger::<F, C::Hasher>::new();
-    for cap in &trace_caps {
-        challenger.observe_cap(cap);
+    for (i, cap) in trace_caps.iter().enumerate() {
+        if KECCAK_TABLES_INDICES.contains(&i) && !use_keccak_tables {
+            // Observe zero merkle caps when skipping Keccak tables.
+            let zero_merkle_cap = cap.flatten().iter().map(|_| F::ZERO).collect::<Vec<F>>();
+            challenger.observe_elements(&zero_merkle_cap);
```

Comment:
Can they not just be ignored? (as in only observe if it's not a keccak table)?

Also, has a similar change been applied on the verifier side (in `get_challenges.rs`)?

Reply:
It is difficult to implement conditionally observe in root circuits, so I choose observe zero caps instead.

I was originally planning to implement it in the next PR (which requires changes on the Plonky2 side, and `get_challenges.rs` is only used for testing). However, I agree with you that we should make this PR complete. I will implement it in this PR instead.